{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import HTSeq\n",
    "import itertools\n",
    "import pysam\n",
    "import collections\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_reader = HTSeq.BAM_Reader(\"408GBM8G-1_S28_L005Aligned.sortedByCoord.out.bam\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = HTSeq.GFF_Reader(\"Homo_sapiens.GRCh38.100.gtf.gz\", end_included=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = HTSeq.GenomicArrayOfSets( \"auto\", stranded=False )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for feature in gtf_file:\n",
    "    if feature.type == \"exon\":\n",
    "        exons[ feature.iv ] += feature.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_table('./first_task/gene_exp.diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>status</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>X:100627107-100639991</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>OK</td>\n",
       "      <td>9.08395</td>\n",
       "      <td>9.69436</td>\n",
       "      <td>0.093827</td>\n",
       "      <td>0.203156</td>\n",
       "      <td>0.82225</td>\n",
       "      <td>0.999515</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>X:100584935-100599885</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>20:50888915-50958555</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>OK</td>\n",
       "      <td>38.40530</td>\n",
       "      <td>40.42290</td>\n",
       "      <td>0.073868</td>\n",
       "      <td>0.090539</td>\n",
       "      <td>0.92635</td>\n",
       "      <td>0.999515</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1:169662006-169894267</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>OK</td>\n",
       "      <td>5.30597</td>\n",
       "      <td>4.26364</td>\n",
       "      <td>-0.315532</td>\n",
       "      <td>-0.548115</td>\n",
       "      <td>0.57965</td>\n",
       "      <td>0.999515</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1:169662006-169894267</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>OK</td>\n",
       "      <td>7.17412</td>\n",
       "      <td>3.74769</td>\n",
       "      <td>-0.936800</td>\n",
       "      <td>-1.154370</td>\n",
       "      <td>0.17215</td>\n",
       "      <td>0.999515</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60656</th>\n",
       "      <td>ENSG00000288600</td>\n",
       "      <td>ENSG00000288600</td>\n",
       "      <td>AL354833.2</td>\n",
       "      <td>13:41488356-41506248</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60657</th>\n",
       "      <td>ENSG00000288601</td>\n",
       "      <td>ENSG00000288601</td>\n",
       "      <td>AC007599.3</td>\n",
       "      <td>16:47144322-47701523</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60658</th>\n",
       "      <td>ENSG00000288602</td>\n",
       "      <td>ENSG00000288602</td>\n",
       "      <td>C8orf44-SGK3</td>\n",
       "      <td>8:66667551-66862022</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60659</th>\n",
       "      <td>ENSG00000288603</td>\n",
       "      <td>ENSG00000288603</td>\n",
       "      <td>SNORA74C-2</td>\n",
       "      <td>10:49818278-50011654</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60660</th>\n",
       "      <td>ENSG00000288604</td>\n",
       "      <td>ENSG00000288604</td>\n",
       "      <td>AC015871.8</td>\n",
       "      <td>15:79833584-79926993</td>\n",
       "      <td>q1</td>\n",
       "      <td>q2</td>\n",
       "      <td>OK</td>\n",
       "      <td>2.44709</td>\n",
       "      <td>1.70492</td>\n",
       "      <td>-0.521364</td>\n",
       "      <td>-0.168190</td>\n",
       "      <td>0.49945</td>\n",
       "      <td>0.999515</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60661 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               test_id          gene_id          gene                  locus  \\\n",
       "0      ENSG00000000003  ENSG00000000003        TSPAN6  X:100627107-100639991   \n",
       "1      ENSG00000000005  ENSG00000000005          TNMD  X:100584935-100599885   \n",
       "2      ENSG00000000419  ENSG00000000419          DPM1   20:50888915-50958555   \n",
       "3      ENSG00000000457  ENSG00000000457         SCYL3  1:169662006-169894267   \n",
       "4      ENSG00000000460  ENSG00000000460      C1orf112  1:169662006-169894267   \n",
       "...                ...              ...           ...                    ...   \n",
       "60656  ENSG00000288600  ENSG00000288600    AL354833.2   13:41488356-41506248   \n",
       "60657  ENSG00000288601  ENSG00000288601    AC007599.3   16:47144322-47701523   \n",
       "60658  ENSG00000288602  ENSG00000288602  C8orf44-SGK3    8:66667551-66862022   \n",
       "60659  ENSG00000288603  ENSG00000288603    SNORA74C-2   10:49818278-50011654   \n",
       "60660  ENSG00000288604  ENSG00000288604    AC015871.8   15:79833584-79926993   \n",
       "\n",
       "      sample_1 sample_2  status   value_1   value_2  log2(fold_change)  \\\n",
       "0           q1       q2      OK   9.08395   9.69436           0.093827   \n",
       "1           q1       q2  NOTEST   0.00000   0.00000           0.000000   \n",
       "2           q1       q2      OK  38.40530  40.42290           0.073868   \n",
       "3           q1       q2      OK   5.30597   4.26364          -0.315532   \n",
       "4           q1       q2      OK   7.17412   3.74769          -0.936800   \n",
       "...        ...      ...     ...       ...       ...                ...   \n",
       "60656       q1       q2  NOTEST   0.00000   0.00000           0.000000   \n",
       "60657       q1       q2  NOTEST   0.00000   0.00000           0.000000   \n",
       "60658       q1       q2  NOTEST   0.00000   0.00000           0.000000   \n",
       "60659       q1       q2  NOTEST   0.00000   0.00000           0.000000   \n",
       "60660       q1       q2      OK   2.44709   1.70492          -0.521364   \n",
       "\n",
       "       test_stat  p_value   q_value significant  \n",
       "0       0.203156  0.82225  0.999515          no  \n",
       "1       0.000000  1.00000  1.000000          no  \n",
       "2       0.090539  0.92635  0.999515          no  \n",
       "3      -0.548115  0.57965  0.999515          no  \n",
       "4      -1.154370  0.17215  0.999515          no  \n",
       "...          ...      ...       ...         ...  \n",
       "60656   0.000000  1.00000  1.000000          no  \n",
       "60657   0.000000  1.00000  1.000000          no  \n",
       "60658   0.000000  1.00000  1.000000          no  \n",
       "60659   0.000000  1.00000  1.000000          no  \n",
       "60660  -0.168190  0.49945  0.999515          no  \n",
       "\n",
       "[60661 rows x 14 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "table = table[['gene', 'locus', 'value_1', 'value_2', 'status', 'log2(fold_change)']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>status</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>X:100627107-100639991</td>\n",
       "      <td>9.08395</td>\n",
       "      <td>9.69436</td>\n",
       "      <td>OK</td>\n",
       "      <td>0.093827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>X:100584935-100599885</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>20:50888915-50958555</td>\n",
       "      <td>38.40530</td>\n",
       "      <td>40.42290</td>\n",
       "      <td>OK</td>\n",
       "      <td>0.073868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1:169662006-169894267</td>\n",
       "      <td>5.30597</td>\n",
       "      <td>4.26364</td>\n",
       "      <td>OK</td>\n",
       "      <td>-0.315532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1:169662006-169894267</td>\n",
       "      <td>7.17412</td>\n",
       "      <td>3.74769</td>\n",
       "      <td>OK</td>\n",
       "      <td>-0.936800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60656</th>\n",
       "      <td>AL354833.2</td>\n",
       "      <td>13:41488356-41506248</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60657</th>\n",
       "      <td>AC007599.3</td>\n",
       "      <td>16:47144322-47701523</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60658</th>\n",
       "      <td>C8orf44-SGK3</td>\n",
       "      <td>8:66667551-66862022</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60659</th>\n",
       "      <td>SNORA74C-2</td>\n",
       "      <td>10:49818278-50011654</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60660</th>\n",
       "      <td>AC015871.8</td>\n",
       "      <td>15:79833584-79926993</td>\n",
       "      <td>2.44709</td>\n",
       "      <td>1.70492</td>\n",
       "      <td>OK</td>\n",
       "      <td>-0.521364</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60661 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene                  locus   value_1   value_2  status  \\\n",
       "0            TSPAN6  X:100627107-100639991   9.08395   9.69436      OK   \n",
       "1              TNMD  X:100584935-100599885   0.00000   0.00000  NOTEST   \n",
       "2              DPM1   20:50888915-50958555  38.40530  40.42290      OK   \n",
       "3             SCYL3  1:169662006-169894267   5.30597   4.26364      OK   \n",
       "4          C1orf112  1:169662006-169894267   7.17412   3.74769      OK   \n",
       "...             ...                    ...       ...       ...     ...   \n",
       "60656    AL354833.2   13:41488356-41506248   0.00000   0.00000  NOTEST   \n",
       "60657    AC007599.3   16:47144322-47701523   0.00000   0.00000  NOTEST   \n",
       "60658  C8orf44-SGK3    8:66667551-66862022   0.00000   0.00000  NOTEST   \n",
       "60659    SNORA74C-2   10:49818278-50011654   0.00000   0.00000  NOTEST   \n",
       "60660    AC015871.8   15:79833584-79926993   2.44709   1.70492      OK   \n",
       "\n",
       "       log2(fold_change)  \n",
       "0               0.093827  \n",
       "1               0.000000  \n",
       "2               0.073868  \n",
       "3              -0.315532  \n",
       "4              -0.936800  \n",
       "...                  ...  \n",
       "60656           0.000000  \n",
       "60657           0.000000  \n",
       "60658           0.000000  \n",
       "60659           0.000000  \n",
       "60660          -0.521364  \n",
       "\n",
       "[60661 rows x 6 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "table[table['status'] == 'OK'].drop(columns='status').rename(columns={'value_1':'408GBM0G-1_S25_L005', 'value_2':'408GBM8G-1_S28_L005'}).to_excel('./first_task/answer2.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
